### PR TITLE
Feature/hd 4220 last block size

### PR DESCRIPTION
**Pull request by fferrandis**
This pull request introduces significant improvements and refactoring to the buffer matrix management for erasure coding, focusing on more efficient handling of data fragments, especially for cases where the last subgroup (stripe) is not a full block. The changes include a new buffer format, more flexible offset calculations, improved test coverage, and a new C header for backend interoperability.

**Key changes:**

### Buffer Matrix Refactoring and New Format Support

* Introduced a new `BufferInfo` struct to encapsulate buffer metadata, separated from the main `BufferMatrix` data structure, and refactored related logic for clarity and maintainability. Added methods to support both old and new buffer layouts, with explicit `UseNewFormat` and `UseOldFormat` setters and corresponding offset calculation logic. (`buffer.go` [[1]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L8-R97) [[2]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R112-L92) [[3]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R165-R242) [[4]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L125-R251) [[5]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R299-R305) [[6]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R317-R341)

* Implemented logic to handle the last subgroup differently when the data size is not a multiple of the block size, ensuring correct data alignment and minimizing wasted space in the last fragments. (`buffer.go` [[1]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R165-R242) [[2]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R299-R305) [[3]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R317-R341)

### Expanded and Improved Testing

* Removed outdated or redundant tests and benchmarks, and added comprehensive new tests to cover the new buffer format, offset calculations, subgroup size computations, and block mapping. This ensures correctness of the new logic and robust coverage for edge cases. (`buffer_test.go` [[1]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L74-L116) [[2]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L134-R93) [[3]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L163-R281)

### Backend Interoperability

* Added a new C header file `backend.h` defining the `encode_chunk_context` structure and a suite of utility functions to facilitate interaction between Go and C code for erasure coding backends. (`backend.h` [backend.hR1-R70](diffhunk://#diff-f142809213f44ad3f9647b0664a4272e61cd1096c051006621c94cc1c628ea47R1-R70))

### Minor Improvements

* Added a `// nolint:gosec` comment to suppress a linter warning for pointer arithmetic in `memalign.go`. (`memalign.go` [memalign.goR9](diffhunk://#diff-97c9d06b50cbe1f96309c4f4a8ada847ea6a372632944fc37a8eb74a49b2e574R9))

These changes collectively modernize the buffer management for erasure coding, improve performance for non-aligned data sizes, and lay groundwork for further backend integration.

**References:** [[1]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L8-R97) [[2]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R112-L92) [[3]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R165-R242) [[4]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4L125-R251) [[5]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R299-R305) [[6]](diffhunk://#diff-ccb968ee74b22f49cb68ca379bc4c3bad1a51b239afffdfd04711c20eb22c6c4R317-R341) [[7]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L74-L116) [[8]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L134-R93) [[9]](diffhunk://#diff-6dfe64bb6fddc99ffec7c6fbdcd1bac98818e510c9310ced6ec544ce2597e1d6L163-R281) [[10]](diffhunk://#diff-f142809213f44ad3f9647b0664a4272e61cd1096c051006621c94cc1c628ea47R1-R70) [[11]](diffhunk://#diff-97c9d06b50cbe1f96309c4f4a8ada847ea6a372632944fc37a8eb74a49b2e574R9)

### CODE REVIEW

File: backend.c [hunk 1]
```diff
@@ -0,0 +1,408 @@
+#include <liberasurecode/erasurecode.h>
```
Comment (fferrandis) [inline backend.c:1]: this is only a move from code that was directly inline in GO file to a proper dedicated C file.

File: backend.go [hunk 2]
```diff
@@ -591,25 +185,13 @@ func InitBackend(params Params) (Backend, error) {
 	} else {
 		backend.pool = &pool{
 			p: sync.Pool{
-				New: func() interface{} {
+				New: func() any {
 					return bytes.NewBuffer(make([]byte, params.MaxBlockSize))
 				}},
 			max: params.MaxBlockSize,
 		}
 	}
```
Comment (fferrandis) [inline backend.go:194]: as we dont use jerasure, this is removed. (this new feature can't work with jerasure anyway)

File: backend.go [hunk 3]
```diff
@@ -684,7 +274,15 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
 
 	for i := 0; i < int(ctx.number_of_subgroup); i++ {
 		go func(nth int) {
```
Comment (fferrandis) [inline backend.go:282]: this is the main diff of the encode part. We make a distinction when it comes to the last lines/stripes. All the stripes are supposed to be "chunksize" long, but not the last one because of the new implementation

File: backend.go [hunk 4]
```diff
@@ -697,7 +295,7 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
 		return &EncodeData{nil, func() {
 				C.my_liberasurecode_encode_buffermatrix_cleanup(
 					backend.libecDesc, C.size_t(ctx.frags_len), ctx.datas, ctx.codings)
-			}},
+			}, totLen},
```
Comment (fferrandis) [inline backend.go:304]: we store the real len of the slice. We could also reslice it to this len, but it was easier from the debug point of view to keep the slice with the same slice regardless of the usage of the new format, and store the real len alongside.

File: backend.go [hunk 5]
```diff
@@ -716,57 +314,7 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
 		runtime.KeepAlive(bm)
 		C.my_liberasurecode_encode_buffermatrix_cleanup(
 			backend.libecDesc, C.size_t(ctx.frags_len), ctx.datas, ctx.codings)
-	}}, nil
-}
-
-// EncodeMatrix encodes data in small subpart of chunkSize bytes
```
Comment (fferrandis) [inline backend.go:722]: EncodeMatrix isn't used in hdcontroller. So all the tests using it has been updated to replace the usage of EncodeMatrix by EncodeMatrixWithBufferMatrix

File: backend.go [hunk 6]
```diff
@@ -972,8 +519,9 @@ func (backend *Backend) DecodeMatrix(frags []ValidatedFragment, pieceSize int) (
 	var totLen int64
 	for i := 0; i < chunkInfo.NrChunk; i++ {
 		vect := make([][]byte, len(frags))
```
Comment (fferrandis) [inline backend.go:537]: when decoding stripes, we can't assume that all stripes are "chunksized long". When using the new format, the last stripe of a fragment is often smaller (that is the prupose of this PR and this new format). The nextbound is computed accordingly.

File: backend.go [hunk 7]
```diff
@@ -1050,41 +598,69 @@ type RangeMatrix struct {
  *     p4 [-[*]- -]
  *
  */
```
Comment (fferrandis) [inline backend.go:619]: this is mainly variable renaming to help to understand what we are manipulating in the matrix. That why we have lines, celles and columns.

File: backend.go [hunk 8]
```diff
@@ -1050,41 +598,69 @@ type RangeMatrix struct {
...
+	nrCellsToRead := (idxEnd + 1 - idxStart)
+	dataOffset := idxStart * cellDataSize
 
-	fragFirstIncl := pieceStartIncl % backend.K
-	fragCount := (pieceEndIncl + 1 - pieceStartIncl)
-	dataOffset := pieceStartIncl * pieceSize
+	totalLines := (maxDataSize + lineSize - 1) / lineSize
+	isLastStripe := (lineEnd == totalLines-1)
 
 	/* When wrapping around, we read the full groups. */
```
Comment (fferrandis) [inline backend.go:672]: this is the main change of GetRangeMatrix : we are going to read a full group (ie a full line of stripes) in case of two condition : 
- first one already existed
- the second one is new and is true when data is in the last stripe line. 
As we cannot guess how the data is stored (using the old format or the new one) we chose to read all the stripes and then, based on the stripes header, get properly the data

**Review by fferrandis — COMMENTED** [review]
some comments to help the review

File: backend.go [hunk 9]
```diff
@@ -637,13 +660,13 @@ func (backend *Backend) GetRangeMatrix(startIncl, endIncl, pieceSize, fragSize i
 	linearizedRangeStartIncl := startIncl - dataOffset
 
 	/* Decoding always works on a group boundary. */
```
Comment (liaud) [inline backend.go]: You may want to update the comment to not mention "group" but "column" given the variable names.

File: backend_test.go [hunk 10]
```diff
@@ -1242,3 +1319,314 @@ func TestRangeMatrix(t *testing.T) {
...
+
+	buf := make([]byte, size)
+	for i := range buf {
+		buf[i] = byte('A' + i%26)
+	}
+
+	bm := NewBufferMatrix(chunkSize, len(buf), backend.K)
+	if useNewFormat {
+		bm.UseNewFormat()
+	}
```
Comment (benoit-a) [inline backend_test.go:1466]: Since new format is default, should we consider adding an `else { /* use old format */ }`

Comment (fferrandis) [reply backend_test.go:1466]: yes, nice catch

**Review by benoit-a — APPROVED** [review]

File: backend.c [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+bool check_matrix_fragment(char *frag, int frag_len, int piecesize) {
+  size_t offset = 0;
+
+  while (offset < frag_len) {
+    if (is_invalid_fragment_header((fragment_header_t *)&frag[offset])) {
+      return false;
+    }
+    offset += piecesize + getHeaderSize();
+  }
+
```
[the remark this answers is not in the document]
Comment (liaud) [reply backend.c:112]: Maybe an alternative would be to check that the remaining has just enough data for the header before calling `is_invalid_fragment_header` in the loop above, then check that the header returned size valid w.r.t to the provided buffer.

File: backend.go [hunk 12]
```diff
@@ -684,7 +280,15 @@ func (backend *Backend) EncodeMatrixWithBufferMatrix(bm *BufferMatrix, chunkSize
 
 	for i := 0; i < int(ctx.number_of_subgroup); i++ {
 		go func(nth int) {
-			r := C.encode_chunk_buffermatrix(backend.libecDesc, pData, pDataLen, nbFrags, &ctx, C.int(nth))
+			fragLen := C.size_t(chunkSize)
+			// last subgroup has a different size
+			if i == int(ctx.number_of_subgroup)-1 {
```
[the remark this answers is not in the document]
Comment (liaud) [reply backend.go]: This suggestion looks correct. I guess it works by "chance" because the last goroutine will have the correct shared `i`.

File: backend_test.go [hunk 13]
```diff
@@ -813,26 +793,147 @@ func TestEncodeM(t *testing.T) {
...
+			found := decoded.Data[rangeM.LinearizedRangeStartIncl:linearizedRangeEndExcl]
+
+			require.True(t, bytes.Equal(expected, found))
+
+			frags2 := make([][]byte, 0)
+			for i := 0; i < backend.K+backend.M; i++ {
+				frags2 = append(frags2, encoded.Data[i][rangeM.InFragRangeStartIncl:rangeM.InFragRangeEndExcl])
+			}
+
+			// now do the same test, but this time, insted of linearizing, we are going to reconstruct the stripes
```
Comment (liaud) [inline backend_test.go]:
```suggestion
			// now do the same test, but this time, instead of linearizing, we are going to reconstruct the stripes
```

**Review by liaud — APPROVED** [review]
